torchgen/dest/dispatchless.py

```diff
@@ -0,0 +1,265 @@
+    Variant,
+)
+from torchgen.utils import assert_never
+
+# necessary information for an operation marked as dispatch-less composite.
+@dataclass(frozen=True)
+class CompositeInfo:
+    # the actual NativeFunction that corresponds to the OperatorName.
+    f: NativeFunction
+
+    # a list of dependencies.
+    # for each dependency, we also save its NativeFunctionsGroup (if there
+    # is one) because of how structured kernels are generated.
+    # if a NativeFunction has no BackendMetadata, it can still be registered
+    # to a dispatch key if it's structured (i.e. its out function is structured).
+    dependencies: List[Tuple[NativeFunction, Optional[NativeFunctionsGroup]]]
+
+
+# Dependency information for dispatch-less composite kernels.
+# Maps the overloaded operator name to a list of operations it depends on.
+# We need to store both the 'NativeFunction' it depends on, as well as the
+# 'NativeFunctionsGroup' it belongs to in order to find out whether the
+# operation has a kernel registered for a given dispatch key.
+CompositeGraph = Dict[OperatorName, CompositeInfo]
+
+
+def get_native_and_group_pair(
+    g: Union[NativeFunction, NativeFunctionsGroup]
+) -> List[Tuple[NativeFunction, Optional[NativeFunctionsGroup]]]:
+    if isinstance(g, NativeFunction):
+        fs = [g]
+        group = None
+    elif isinstance(g, NativeFunctionsGroup):
+        fs = list(g.functions())
+        group = g
+    else:
+        assert_never(g)
+    return [(f, group) for f in fs]
+
+
+def get_graph(
+    grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
+    b: BackendIndex,
+) -> CompositeGraph:
+    opname_to_native_map = {
+        f.func.name: (f, group)
+        for g in grouped_native_functions
+        for f, group in get_native_and_group_pair(g)
+    }
+
+    return {
+        f.func.name: CompositeInfo(
+            f,
+            sorted(
+                (opname_to_native_map[opname] for opname in f.composite[0]), key=str
+            ),
+        )
+        for g in grouped_native_functions
+        for f, _ in get_native_and_group_pair(g)
+        if len(f.composite[0]) > 0 and b.dispatch_key in f.composite[1]
+    }
+
+
+# convenience function for checking if a NativeFunction is registered to a
+# dispatch key.
+def has_registered_kernel(
+    index: BackendIndex, f: NativeFunction, g: Optional[NativeFunctionsGroup]
+) -> bool:
+    if g is not None and g.structured:
+        # if f is structured, then it will be registered for both Meta and
+        # CompositeExpicitAutograd (both not in STRUCTURED_DISPATCH_KEYS).
+        cond_meta = index.dispatch_key == DispatchKey.Meta
+
+        # structured kernel generation does not register out functions for
+        # CompositeExplicitAutograd.
+        cond_composite = (
+            index.dispatch_key == DispatchKey.CompositeExplicitAutograd
+            and not f.func.is_out_fn()
+        )
+        return index.has_kernel(g) or cond_meta or cond_composite
```

Comment:
Just thinking out loud, but I wonder if it'd be worth at some point getting `index.has_kernel(f)` to return True properly if you're a `structured_delegate` who's `out=` variant has a valid kernel. If we do that, then I don't think you'd need to do all that work plumbing `Union[NativeFunction, NativeFunctionsGroup]` through the `CompositeGraph`.

No action needed though, I think the extra plumbing doesn't seem too bad.

I think the only reason you need to all the logic to plumb `Tuple[NativeFunction, NativeFunctionsGroup]` objects through the `CompositeGraph` is basically because of this

Reply:
Not sure if we need this information anywhere else... But, sure would be nice to avoid dispatcher trips if we could.
It's not only because of `structured_delegate` kernels, but also `Meta` and `CompositeExplicitAutograd` dispatch keys. Those are not in `STRUCTURED_DISPATCH_KEYS`, i.e. there's no `BackendMetadata` created for them.